### ISSUE TITLE
Quirky behaviour with predicates

### ISSUE THREAD

**Issue by william-chu-github** [shapely/shapely#1551]
I have noticed some behaviour which does not seem to match with the Shapely documents. I don't know if it is a bug, if the documentation is wrong, or I'm misunderstanding something about the package.

[The manual says: The interior, boundary, and exterior sets of a feature are mutually exclusive and their union coincides with the entire plane.
](https://shapely.readthedocs.io/en/stable/manual.html#spatial-data-model)

[It additionally says, of the `.touches()` method: Returns `True` if the objects have at least one point in common and their interiors do not intersect with any part of the other.](https://shapely.readthedocs.io/en/stable/manual.html#object.touches)

I'm going to set up three linestrings:

```
>>> Square = shapely.geometry.LineString([(0, 0), (0, 2), (2, 2), (2, 0), (0, 0)])
>>> Triangle = shapely.geometry.LineString([(1, 0), (0.5, 1), (1.5, 1), (1, 0)])
>>> TriangleMinusClosure = shapely.geometry.LineString([(1, 0), (0.5, 1), (1.5, 1)])
```
The triangle intersects the square in the middle of the bottommost segment. The `TriangleMinusClosure` is simply
two legs of the triangle without the closing segment; it still intersects the square in the middle of the bottommost segment.

```
>>> Int1 = Square.intersection(Triangle)
>>> Int2 = Square.intersection(TriangleMinusClosure)
>>> type(Int1)
<class 'shapely.geometry.point.Point'>
>>> type(Int2)
<class 'shapely.geometry.point.Point'>
>>> Int1 == Int2
True
```

So far, so good. It falls apart when I try to apply `.touches()`:

```
>>> Triangle.touches(Square)
False
>>> TriangleMinusClosure.touches(Square)
True
```

Both `Triangle` and `TriangleMinusClosure` have at least one point in common with `Square`. To fulfill the other half of the `.touches()` requirement, they must not have interior points intersecting with `Square`. Since the interior, boundary, and exterior of a given object are mutually exclusive, and the boundary of a linestring is its two endpoints, I would expect both to touch the square--their only interaction with the square is a boundary point.

I see that the closed linestring is reporting an empty boundary, which would imply (?) that the interior is is everything on the triangle, which would explain the `False` touch result.

```
>>> list(Triangle.boundary)
[]
>>> list(TriangleMinusClosure.boundary)
[<shapely.geometry.point.Point object at 0x00000223AC54A880>, <shapely.geometry.point.Point object at 0x00000223AC54A7F0>]
```

The DE-9IM code does seem to indicate a pointwise intersection of the triangle's interior with the square's interior:
```
>>> Triangle.relate(Square)
'0F1FFF1F2'
>>> TriangleMinusClosure.relate(Square)
'FF10F01F2'
```

Windows 10 Home 21H1

Shapely version 1.8.1.post1 via pip

**Comment by mwtoews** [comment]
Note that this is identical to what you see with JTS given these WKT for convenience:

- Square: `LINESTRING (0 0, 0 2, 2 2, 2 0, 0 0)`
- Triangle: `LINESTRING (1 0, 0.5 1, 1.5 1, 1 0)`
- TriangleMinusClosure: `LINESTRING (1 0, 0.5 1, 1.5 1)`

I can't say if this is expected or not, perhaps @dr-jts could see if this expected.

**Comment by jorisvandenbossche** [comment]
> I see that the closed linestring is reporting an empty boundary, which would imply (?) that the interior is is everything on the triangle, which would explain the `False` touch result.

This is a bit strange (and indeed from having no boundary, it follows that they do not touch). For a LineString, the boundary is the two end points, so I personally wouldn't expect it to matter whether the linestring is closed / the two endpoints are the same. However, there is some explicit logic for this in GEOS:

**Comment by jorisvandenbossche** [comment]
Another interesting quote from the method that checks the "closed endpoint on boundary" (https://github.com/libgeos/geos/blob/df8b93c4b274fe6d6fda91e4c82f339f83f31b90/include/geos/algorithm/BoundaryNodeRule.h#L71-L80):

> The Mod-2 Boundary Node Rule (which is the rule specified
> in the OGC SFS).
> 
> A BoundaryNodeRule specifies that points are in the boundary of
> a lineal geometry iff the point lies on the boundary of an odd number
> of components. Under this rule LinearRings and closed LineStrings have
> an empty boundary.
>    
> This is the rule specified by the OGC SFS, and is the default
> rule used in JTS.

**Comment by william-chu-github** [comment]
So it looks like this may be working as intended, even if it is counterintuitive? I do see a reference in the docs about linear rings having an empty boundary, but no mention of closed linestrings (although I guess this is implied by the linear ring statement).

> Another interesting quote from the method that checks the "closed endpoint on boundary"

I don't know enough C to comment intelligently, but the definition [after the one you linked](https://github.com/libgeos/geos/blob/df8b93c4b274fe6d6fda91e4c82f339f83f31b90/include/geos/algorithm/BoundaryNodeRule.h#L85) seems like it indicates an option to allow LinearRings to have a nonempty boundary set:

```
     * A BoundaryNodeRule which specifies that any points which are endpoints
     * of lineal components are in the boundary of the parent geometry. This
     * corresponds to the "intuitive" topological definition of boundary. Under
     * this rule LinearRings have a non-empty boundary (the common endpoint
     * of the underlying LineString).
```

**Comment by jorisvandenbossche** [comment]
Yes, but so that explanation is for a different rule, and the one that is used by default is the "Mod-2" rule that I quoted. 

It seems that the C API of GEOS does expose those rules, and so you can could call `relate` with specifying a different rule. But this isn't exposed in Shapely at the moment, but contributions are certainly welcome (https://github.com/libgeos/geos/blob/df8b93c4b274fe6d6fda91e4c82f339f83f31b90/capi/geos_c.h.in#L1286-L1301 and https://github.com/libgeos/geos/blob/df8b93c4b274fe6d6fda91e4c82f339f83f31b90/capi/geos_c.h.in#L4236-L4250)

(that wouldn't solve it for `touches` itself, but it would allow a workaround with calling `relate` and checking the pattern for "touches" patterns)

**Comment by jorisvandenbossche** [comment]
> Another way to look at this is that the closed triangle is topologically identical to a triangle with its endpoints at one of the other vertices. The only way to make topologically operations consistent with this is to specify that closed curves have no boundaries.

Ah, this is an interesting way to look at it indeed, that makes me understand the reason of the rule.

We should probably document this corner case in the manual about "touches".

---

Something else I am wondering on this topic: our manual says that a Point is considered to have an interior of that single point, and an empty boundary. But if you follow that logic, a Point should never touch another geometry? (since the intersecting point will always lie in the interior) In practice, Points do touch other geometries (eg `shapely.box(0, 0, 1, 1).touches(shapely.Point(0, 0))` gives True), so that is maybe special cases to give more sensible behaviour?

**Comment by dr-jts** [comment]
> Something else I am wondering on this topic: our manual says that a Point is considered to have an interior of that single point, and an empty boundary. But if you follow that logic, a Point should never touch another geometry? (since the intersecting point will always lie in the interior) In practice, Points do touch other geometries (eg `shapely.box(0, 0, 1, 1).touches(shapely.Point(0, 0))` gives True), so that is maybe special cases to give more sensible behaviour?

Here's the definition of Touches from the SFS:
![image](https://user-images.githubusercontent.com/3529053/192598206-ff935da4-f293-4554-9135-cb4bc214d2b8.png)
It requires that the geometries intersect, but their interiors are disjoint.  This is the case in the Polygon/Point case above - the Interior of the Point intersects the Boundary of the Polygon, but not its Interior.
